test/functional/test-intersection-observer-polyfill.js

```diff
@@ -0,0 +1,461 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {IntersectionObserverPolyfill,
+} from '../../src/intersection-observer-polyfill';
+import {layoutRectLtwh} from '../../src/layout-rect';
+import * as sinon from 'sinon';
+
+describe('IntersectionObserverPolyfill', () => {
+  let sandbox;
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  it('getThresholdSlot_', () => {
+    // default threshold
+    let ioInstance = new IntersectionObserverPolyfill(() => {});
+    expect(ioInstance.getThresholdSlot_(0)).to.equal(0);
+    expect(ioInstance.getThresholdSlot_(0.67)).to.equal(14);
+    expect(ioInstance.getThresholdSlot_(0.65)).to.equal(14);
+    expect(ioInstance.getThresholdSlot_(1)).to.equal(21);
+    let options = {threshold: [0.01, 0.25, 0.5, 1]};
+    ioInstance = new IntersectionObserverPolyfill(() => {}, options);
+    ioInstance.threshold_ = [0.01, 0.25, 0.5, 1];
+    expect(ioInstance.getThresholdSlot_(0)).to.equal(0);
+    options = {threshold: [0, 0.25, 0.5, 1]};
+    ioInstance = new IntersectionObserverPolyfill(() => {}, options);
+    expect(ioInstance.getThresholdSlot_(0)).to.equal(0);
+    expect(ioInstance.getThresholdSlot_(0.1)).to.equal(1);
+    expect(ioInstance.getThresholdSlot_(0.4)).to.equal(2);
+    expect(ioInstance.getThresholdSlot_(0.25)).to.equal(2);
+    expect(ioInstance.getThresholdSlot_(0.5)).to.equal(3);
+    expect(ioInstance.getThresholdSlot_(1)).to.equal(4);
+    options = {threshold: [0.5]};
+    ioInstance = new IntersectionObserverPolyfill(() => {}, options);
+    ioInstance.threshold_ = [0.5];
+    expect(ioInstance.getThresholdSlot_(1)).to.equal(1);
+  });
+
+  describe('tick function', () => {
+    let element;
+    element = {
+      isBuilt: () => {return true;},
+    };
+
+    describe('w/o container should get IntersectionChangeEntry when', () => {
+      it('intersect correctly base', done => {
+        sandbox.stub(performance, 'now', () => {
+          return 100;
+        });
+        const ioInstance = new IntersectionObserverPolyfill(change => {
```

Comment:
code might be more readable if you use a spy here.
```
forEach(() => {
  const callbackSpy = sandbox.spy();
  io = new IntersectionObserverPolyfill(spy);
});
```

then in each test case you can check if spy is called with correct param.